src/lambda_function/raw/app.py

```diff
@@ -0,0 +1,337 @@
+    # The first underscore-delimited component of the JSON basename is the datatype
+    data_type = file_basename.split("_")[0]
+    raw_basename = f"{ os.path.splitext(file_basename)[0] }.ndjson.gz"
+    raw_key = os.path.join(
+        raw_key_prefix,
+        f"dataset={data_type}",
+        f"cohort={cohort}",
+        raw_basename,
+    )
+    return raw_key
+
+
+def upload_part(
+    s3_client: boto3.client,
+    body: bytes,
+    bucket: str,
+    key: str,
+    upload_id: str = None,
+    part_number: int = None,
+):
+    """
+    Uploads a part of data to an S3 object as part of a multipart upload.
+
+    If an `upload_id` is not provided, the function initiates a new multipart
+    upload. Each part is identified by its ETag and part number, which are
+    required for the completion the multipart upload.
+
+    Args:
+        s3_client (boto3.client): The Boto3 S3 client used to interact with AWS S3.
+        body (bytes): The data to be uploaded as a part, in bytes.
+        bucket (str): The name of the raw S3 bucket where the object is to be stored.
+        key (str): The S3 key of the object being uploaded.
+        upload_id (str, optional): The ID of an ongoing multipart upload. If not provided,
+            a new multipart upload is initiated.
+        part_number (int, optional): The part number for this chunk of the upload.
+            If not provided, defaults to 1 when initiating a new upload.
+
+    Returns:
+        dict: A dictionary containing the following keys:
+            - part (dict): This dictionary must be included with the other parts
+                in a list when completing the multipart upload and is provided
+                in this format for convenience. A dictionary with keys:
+                    * ETag (str): The ETag of this part.
+                    * part_number (int) The associated part number.
+            - upload_id (str): The ID of the multipart upload.
+            - part_number (int): The part number associated with this upload part.
+
+    Logs:
+        - Logs the start of a new multipart upload if `upload_id` is not provided.
+        - Logs the upload size and S3 destination for each part.
+        - Logs the response from the S3 upload operation, including the ETag.
+
+    Example:
+        # Upload a part to an existing multipart upload
+        s3 = boto3.client('s3')
+        part = upload_part(
+            s3_client=s3,
+            body=b'some data chunk',
+            bucket='my-bucket',
+            key='my-object',
+            upload_id='existing-upload-id',
+            part_number=2
+        )
+        print(part)
+        # Output:
+        # {
+        #   'part':
+        #     {
+        #       'ETag': 'etag-value',
+        #       'PartNumber': 2
+        #     },
+        #   'upload_id': 'existing-upload-id',
+        #   'part_number': 2
+        # }
+
+    Notes:
+        - Parts must be larger than AWS minimum requirements (typically 5 MB),
```

Comment:
Did you need to configure anything to make sure the parts are > 5MB?

Reply:
That's taken care of by `yield_compressed_data`. The `part_threshold` is set to 8 MB by default.